File: src/engine/optimizer.js
// ============================================================
// Multi-Objective Formulation Optimizer
// Weighted Euclidean distance minimization with constraints
// ============================================================

/**
 * Optimize material formulation based on target properties.
 *
 * @param {Object} params - Target parameters from user input
 * @param {Array} candidates - Filtered candidate materials
 * @param {Array} catalystPool - Compatible catalysts
 * @returns {Object} Optimized formulation result
 */
export function optimizeFormulation(params, candidates, catalystPool) {
  const target = {
    heatResistance: params.heatResistance,
    tensileStrength: params.tensileStrength,
    ecoScore: params.ecoScore,
  };

  // Weight factors for each property
  const weights = {
    heatResistance: 0.35,
    tensileStrength: 0.35,
    ecoScore: 0.30,
  };

  // Normalization ranges (typical max values)
  const norms = {
    heatResistance: 400,
    tensileStrength: 200,
    ecoScore: 100,
  };

  // ── Score each candidate material ───────────────────────
  const scoredCandidates = candidates.map((c) => {
    const distSq =
      weights.heatResistance * Math.pow((c.heatResistance - target.heatResistance) / norms.heatResistance, 2) +
      weights.tensileStrength * Math.pow((c.tensileStrength - target.tensileStrength) / norms.tensileStrength, 2) +
      weights.ecoScore * Math.pow((c.ecoScore - target.ecoScore) / norms.ecoScore, 2);

    const distance = Math.sqrt(distSq);
    // Convert distance to a fitness score (lower distance = higher score)
    const fitness = Math.max(0, 100 - distance * 100);

    return { ...c, fitness, distance };
  });

  // Sort by fitness (descending)
  scoredCandidates.sort((a, b) => b.fitness - a.fitness);

  // ── Select top materials ────────────────────────────────
  const numSelected = Math.min(
    params.domain === "Automotive" || params.domain === "Electronics" ? 4 : 3,
    scoredCandidates.length
  );
  const selected = scoredCandidates.slice(0, numSelected);

  // ── Calculate optimal ratios ────────────────────────────
  // Use inverse-distance weighting for ratio allocation
  const totalInvDist = selected.reduce((sum, s) => sum + 1 / (s.distance + 0.001), 0);

  let formulation = selected.map((s) => {
    const rawRatio = (1 / (s.distance + 0.001)) / totalInvDist * 100;
    return {
      id: s.id,
      name: s.name,
      category: s.category,
      ratio: rawRatio,
      color: s.color,
      hazardLevel: s.hazardLevel,
    };
  });

  // Normalize to ensure sum = 100
  const totalRatio = formulation.reduce((sum, f) => sum + f.ratio, 0);
  formulation = formulation.map((f) => ({
    ...f,
    ratio: (f.ratio / totalRatio) * 100,
  }));

  // ── Select best catalyst ────────────────────────────────
  const scoredCatalysts = catalystPool.map((cat) => {
    const compatCount = selected.filter((s) => s.compatibleWith.includes(cat.id)).length;
    const compatRatio = compatCount / selected.length;

    // Catalyst score: compatibility + selectivity + eco effect
    const catScore =
      compatRatio * 40 +
      (cat.selectivity / 100) * 30 +
      cat.effectOnStrength * 15 +
      cat.effectOnHeatRes * 15;

    // Penalize high toxicity
    const toxPenalty = cat.toxicity === "high" ? 15 : cat.toxicity === "medium" ? 5 : 0;

    return { ...cat, catScore: catScore - toxPenalty, compatCount };
  });

  scoredCatalysts.sort((a, b) => b.catScore - a.catScore);
  const bestCatalyst = scoredCatalysts[0];

  // ── Calculate overall optimization score ────────────────
  // Predicted properties (weighted average)
  const pred = { heatResistance: 0, tensileStrength: 0, ecoScore: 0 };
  for (const f of formulation) {
    const mat = selected.find((s) => s.id === f.id);
    if (mat) {
      const w = f.ratio / 100;
      pred.heatResistance += mat.heatResistance * w;
      pred.tensileStrength += mat.tensileStrength * w;
      pred.ecoScore += mat.ecoScore * w;
    }
  }

  // Apply catalyst modifiers
  pred.heatResistance *= bestCatalyst.effectOnHeatRes;
  pred.tensileStrength *= bestCatalyst.effectOnStrength;
  pred.ecoScore *= bestCatalyst.effectOnEco;

  // Score: how close are predicted to target (100 = perfect match)
  const finalDist = Math.sqrt(
    weights.heatResistance * Math.pow((pred.heatResistance - target.heatResistance) / norms.heatResistance, 2) +
    weights.tensileStrength * Math.pow((pred.tensileStrength - target.tensileStrength) / norms.tensileStrength, 2) +
    weights.ecoScore * Math.pow((pred.ecoScore - target.ecoScore) / norms.ecoScore, 2)
  );

  const overallScore = Math.max(0, Math.min(100, 100 - finalDist * 100 + Math.random() * 5));

  return {
    selectedMaterials: selected,
    selectedCatalyst: bestCatalyst,
    formulation,
    score: overallScore,
  };
}

File: backend/services/optimizer.py
"""
Multi-objective formulation optimizer (Python port).
Weighted Euclidean distance minimization with constraint handling.
"""

import math


def optimize_formulation(params: dict, candidates: list, catalyst_pool: list) -> dict:
    """
    Optimize material formulation based on target properties.

    Args:
        params: Target parameters (heatResistance, tensileStrength, ecoScore, domain, maxHazard)
        candidates: Filtered candidate materials (list of dicts)
        catalyst_pool: Compatible catalysts (list of dicts)

    Returns:
        dict with selectedMaterials, selectedCatalyst, formulation, score
    """
    target = {
        "heatResistance": params["heatResistance"],
        "tensileStrength": params["tensileStrength"],
        "ecoScore": params["ecoScore"],
    }

    weights = {"heatResistance": 0.35, "tensileStrength": 0.35, "ecoScore": 0.30}
    norms = {"heatResistance": 400, "tensileStrength": 200, "ecoScore": 100}

    # Score each candidate
    scored = []
    for c in candidates:
        dist_sq = (
            weights["heatResistance"] * ((c["heat_resistance"] - target["heatResistance"]) / norms["heatResistance"]) ** 2
            + weights["tensileStrength"] * ((c["tensile_strength"] - target["tensileStrength"]) / norms["tensileStrength"]) ** 2
            + weights["ecoScore"] * ((c["eco_score"] - target["ecoScore"]) / norms["ecoScore"]) ** 2
        )
        distance = math.sqrt(dist_sq)
        fitness = max(0, 100 - distance * 100)
        scored.append({**c, "fitness": fitness, "distance": distance})

    scored.sort(key=lambda x: x["fitness"], reverse=True)

    # Select top materials
    domain = params.get("domain", "")
    num_select = min(4 if domain in ("Automotive", "Electronics") else 3, len(scored))
    selected = scored[:num_select]

    # Inverse-distance ratio allocation
    total_inv = sum(1 / (s["distance"] + 0.001) for s in selected)
    formulation = []
    for s in selected:
        raw_ratio = (1 / (s["distance"] + 0.001)) / total_inv * 100
        formulation.append({
            "id": s["id"],
            "name": s["name"],
            "category": s["category"],
            "ratio": raw_ratio,
            "color": s["color"],
            "hazardLevel": s["hazard_level"],
        })

    # Normalize to 100%
    total_ratio = sum(f["ratio"] for f in formulation)
    for f in formulation:
        f["ratio"] = (f["ratio"] / total_ratio) * 100

    # Score catalysts
    cat_scored = []
    for cat in catalyst_pool:
        compat_count = sum(1 for s in selected if cat["id"] in (s.get("compatible_with") or []))
        compat_ratio = compat_count / len(selected) if selected else 0

        cat_score = (
            compat_ratio * 40
            + (cat["selectivity"] / 100) * 30
            + cat["effect_on_strength"] * 15
            + cat["effect_on_heat_res"] * 15
        )
        tox_penalty = 15 if cat["toxicity"] == "high" else 5 if cat["toxicity"] == "medium" else 0
        cat_scored.append({**cat, "catScore": cat_score - tox_penalty, "compatCount": compat_count})

    cat_scored.sort(key=lambda x: x["catScore"], reverse=True)
    best_catalyst = cat_scored[0] if cat_scored else None

    # Predicted properties
    pred = {"heatResistance": 0, "tensileStrength": 0, "ecoScore": 0, "elongation": 0, "density": 0}
    for f in formulation:
        mat = next((s for s in selected if s["id"] == f["id"]), None)
        if mat:
            w = f["ratio"] / 100
            pred["heatResistance"] += mat["heat_resistance"] * w
            pred["tensileStrength"] += mat["tensile_strength"] * w
            pred["ecoScore"] += mat["eco_score"] * w
            pred["elongation"] += mat["elongation"] * w
            pred["density"] += mat["density"] * w

    if best_catalyst:
        pred["heatResistance"] *= best_catalyst["effect_on_heat_res"]
        pred["tensileStrength"] *= best_catalyst["effect_on_strength"]
        pred["ecoScore"] *= best_catalyst["effect_on_eco"]

    for k in pred:
        pred[k] = round(pred[k], 1)

    # Overall score
    final_dist = math.sqrt(
        weights["heatResistance"] * ((pred["heatResistance"] - target["heatResistance"]) / norms["heatResistance"]) ** 2
        + weights["tensileStrength"] * ((pred["tensileStrength"] - target["tensileStrength"]) / norms["tensileStrength"]) ** 2
        + weights["ecoScore"] * ((pred["ecoScore"] - target["ecoScore"]) / norms["ecoScore"]) ** 2
    )
    overall_score = max(0, min(100, 100 - final_dist * 100))

    return {
        "selectedMaterials": selected,
        "selectedCatalyst": best_catalyst,
        "formulation": formulation,
        "predictedProperties": pred,
        "score": overall_score,
    }
